crates/pathfinder/src/sync/p2p.rs

```diff
@@ -131,6 +144,82 @@ impl Sync {
 
         Ok(())
     }
+
+    async fn sync_transactions(&self) -> anyhow::Result<()> {
+        const NUM_CONCURRENT_TASKS: usize = 10;
+        let mut tasks = JoinSet::new();
+        let blocks_without_transactions =
+            transactions::blocks_without_transactions(self.storage.clone()).fuse();
+        pin_mut!(blocks_without_transactions);
+        for _ in 0..NUM_CONCURRENT_TASKS {
+            let Some(block) = blocks_without_transactions.next().await else {
+                break;
+            };
+            let block = block?;
+            tasks.spawn(sync_transactions_for_block(
+                self.p2p.clone(),
+                self.storage.clone(),
+                block,
+            ));
+        }
+        while let Some(result) = tasks.join_next().await {
+            result??;
+            if let Some(block) = blocks_without_transactions.next().await {
+                let block = block?;
+                tasks.spawn(sync_transactions_for_block(
+                    self.p2p.clone(),
+                    self.storage.clone(),
+                    block,
+                ));
```

Comment:
This cannot be done concurrently as sqlite can only have one writer at a time. You will need to separate inserting from processing if you wish to have concurrency in processing.

i.e. you can only have a single writer task / thread / stage.

If we only have a single writer, we can also enforce that we only write data in-order i.e. block `[0,1,2,3,4...]` get stored. This gives a nice property in that we don't need to check every block for missing transactions. We can instead find the first block with missing transactions as a starting point.

As rough pseudo-code, with different stages using rust streams:

```rust
let start = db.oldest_block_with_no_transactions().await?;

p2p.transaction_stream(start)  // Vec<Transactions>
   .enumerate()                // (idx, Vec<_>)
   .get_header(start)          // (header, Vec<_>)
   .verify()
   .insert()
   .await;
```

Some additional thoughts. We have several places in which we can apply concurrency.

1. Transaction streaming. This would be internal to the p2p client and probably opaque to this code.
2. Fetching headers - probably not worth it, but we can make this concurrent by using `.buffer(N)` on top of it.
3. Verification has two ways of concurrency. 
  1. Process blocks concurrently (`.buffer(N)`).
  2. Calculate commitment and transaction hashes concurrently. I think this happens already now.

Reply:
One can also split transaction hash calculation from the commitment calculation which might make performance issues easier to diagnose. Although probably unecessary here.